src/main/scala/com/twitter/finagle/postgres/messages/Packet.scala

```diff
@@ -13,10 +12,11 @@ object Packet {
  *
  * Converts content into byte format expected by Postgres.
  */
-case class Packet(code: Option[Char], length: Int, content: ChannelBuffer, inSslNegotation: Boolean = false) {
-  def encode(): ChannelBuffer = {
-    val result = ChannelBuffers.dynamicBuffer()
-    code.map { c =>
+case class Packet(code: Option[Char], length: Int, content: ByteBuf, inSslNegotation: Boolean = false) {
```

Comment:
Now that this holds a `ByteBuf`, which is reference counted, you need to handle it as so. I think these are all intermediates so I suspect all you need to do is make sure you release the underlying `content` once you're done with them. That likely means at the end of the `PgClientChannelHandler.encode` method and at the end of the `BackendMessageParser. parse` method. It seems that you're also using them in the `DataRow` `BackendMessage`, and I don't know where those go. If they escape the netty pipeline I'd suggest copying the data into an on-heap datatype like a `Buf` or an `Array[Byte]`, whichever suits your needs better.

Reply:
Hmm... well, a `Packet`'s content should always end up being delivered to Netty's `out`. But I'm not sure about `DataRow`.  I'll see about replacing its use with `Buf`.

Reply:
It looks like `DataRow` uses only unretained slices of incoming buffers from Netty, which should (from what I understand) be freed appropriately by Netty. So I *think* we should be good on that front. I suppose the only way to know would be to run some long-lived profiling and watch the heap size (assuming the default pooled allocator uses on-heap buffers?)

Reply:
The `DataRow` escapes to the user, doesn't it? That means that that data needs to be released by the user or you risk a memory leak. And since each of its elements are unretained slices of the `Packet`'s `ByteBuf`, there isn't a unique notion of ownership.